app/services/idempotency/headers_validator.rb

```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+module Idempotency
+  class HeadersValidator
+    include ActiveModel::Validations
+
+    attr_reader :idempotency_key
+
+    validates :idempotency_key, presence: true
+    # a relatively relaxed regex to allow for any uuid version
+    validates_format_of :idempotency_key, with: /\A[0-9A-F]{8}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{12}\Z/i
+
+    def initialize(headers)
+      # NB: we need to do a case-insensitive match for IDEMPOTENCY_KEY; rails does not do this automatically if the key contains an underscore
```

Comment:
Hmm.. not convinced this is a great idea. Headers are case sensitive by design and we don't do this for any of the other request headers. Would be better for suppliers to include the correct header 🤔

Reply:
..and potentially raise an invalid request error if it's not present when we expect it.

Reply:
Hi @smoothcontract headers are supposed to be case insensitive; see [section 4.2 of RFC 2616](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2). So I think it is important that we support this.

Reply:
I think rails follows that RFC so whatever we're doing elsewhere should be updated. Maybe separately, though

Reply:
Been testing this locally and Rails/Rack does a few strange things. First, any underscores in the header names are converted to dashes, and then using the `[]` method against `request.headers` is already case insensitive. So you can replace the line below with `@idempotency_key = headers['Idempotency_key']`.

Reply:
Found the docs for this: https://api.rubyonrails.org/classes/ActionDispatch/Http/Headers.html

Reply:
now a case-insensitive header, `Idempotency-Key`